### PR TITLE
[Analysis] Add Scalable field in MemoryLocation.h

### PR DESCRIPTION
**Pull request by harviniriawan**
This is the first of a series of patch to improve Alias Analysis on
  Scalable quantities.
  Keep Scalable information from TypeSize which
  will be used in Alias Analysis.

### CODE REVIEW

**Comment by davemgreen** [comment]
A few things I've noticed, from a look around:
 - The checks in BasicAA may not be high enough.
 - LocationSize::unionWith could do with a safe fallback for scalable vectors. It appears to be used from AST.
 - getMinimalExtentFrom doesn't look correct for scalable vectors
 - isObjectSize too in BasicAA, along with constantOffsetHeuristic although I'm not sure that is reachable at the moment for scalable values.
There may be more but it is worth checking through the uses of LocationSize to make sure non of them look like they would cause problems for Scalable sizes.

File: llvm/include/llvm/Analysis/MemoryLocation.h [hunk 1]
```diff
@@ -77,6 +77,7 @@ class LocationSize {
   };
 
   uint64_t Value;
+  bool Scalable = false;
```
Comment (nikic) [inline llvm/include/llvm/Analysis/MemoryLocation.h:80]: This needs to be packed into Value, otherwise you're making LocationSize two times larger.

**Comment by nikic** [comment]
I think if we support scalable LocationSize, we probably need to make getValue() return a TypeSize instead of plain uint64_t. That will force us to properly handle the scalable case everywhere (getting an assertion failure if we don't, rather than introducing a silent miscompile).

File: llvm/include/llvm/Analysis/MemoryLocation.h [hunk 2]
```diff
@@ -292,6 +307,10 @@ class MemoryLocation {
                           const AAMDNodes &AATags = AAMDNodes())
       : Ptr(Ptr), Size(Size), AATags(AATags) {}
 
+  explicit MemoryLocation(const Value *Ptr, uint64_t Size,
+                          const AAMDNodes &AATags = AAMDNodes())
+      : Ptr(Ptr), Size(Size, false), AATags(AATags) {}
```
Comment (nikic) [inline llvm/include/llvm/Analysis/MemoryLocation.h:312]: Why does this need an explicit ctor now?

File: llvm/lib/Analysis/BasicAliasAnalysis.cpp [hunk 3]
```diff
@@ -110,13 +110,13 @@ static uint64_t getObjectSize(const Value *V, const DataLayout &DL,
   Opts.RoundToAlign = RoundToAlign;
   Opts.NullIsUnknownSize = NullIsValidLoc;
   if (getObjectSize(V, Size, DL, &TLI, Opts))
-    return Size;
-  return MemoryLocation::UnknownSize;
+    return LocationSize(Size, DL.getTypeAllocSize(V->getType()).isScalable());
```
Comment (nikic) [inline llvm/lib/Analysis/BasicAliasAnalysis.cpp:113]: `getObjectSize()` doesn't support scalable vectors, so this change doesn't really make sense to me (same in DSE).

If it's helpful using code to return LocationSize here that's fine, but we shouldn't use this misleading `isScalable()` check here.

File: llvm/lib/CodeGen/StackProtector.cpp [hunk 4]
```diff
@@ -176,8 +176,9 @@ static bool HasAddressTaken(const Instruction *AI, TypeSize AllocSize,
     const auto *I = cast<Instruction>(U);
     // If this instruction accesses memory make sure it doesn't access beyond
     // the bounds of the allocated object.
+    // TODO: TypeSize::getFixed should be modified to adapt to scalable vectors
```
Comment (nikic) [inline llvm/lib/CodeGen/StackProtector.cpp:179]: Isn't this just a matter of replacing `TypeSize::getFixed(MemLoc->Size.getValue())` with plain `MemLoc->Size.getValue()` (which already is a TypeSize now)?

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp [hunk 5]
```diff
@@ -984,9 +985,15 @@ struct DSEState {
...
 
-    const uint64_t KillingSize = KillingLocSize.getValue();
-    const uint64_t DeadSize = DeadLoc.Size.getValue();
+    const TypeSize KillingSize = KillingLocSize.getValue();
+    const TypeSize DeadSize = DeadLoc.Size.getValue();
+    const bool AnyScalable =
+        DeadSize.isScalable() || KillingLocSize.isScalable();
 
+    // TODO: Remove AnyScalable constraint once alias analysis fully support
+    // scalable quantities
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:994]: For the purposes of just this code, is AA support needed, rather than adjusting the size comparisons to use the proper helpers. E.g. in the MustAlias case (which can be determined without extra AA support), we can replace `KillingSize >= DeadSize` with `TypeSize::isKnownGE(KillingSize, DeadSize)` or something?

**Review by nikic — COMMENTED** [review]
General direction looks reasonable to me now.

File: llvm/lib/CodeGen/StackProtector.cpp (hunk 4 above)
Comment (harviniriawan) [reply llvm/lib/CodeGen/StackProtector.cpp:179]: yes agreed

File: llvm/lib/Analysis/BasicAliasAnalysis.cpp (hunk 3 above)
Comment (harviniriawan) [reply llvm/lib/Analysis/BasicAliasAnalysis.cpp:113]: I see, ok

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp (hunk 5 above)
Comment (nikic) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:994]: Can we skip only this code path for scalable vectors then?

**Comment by harviniriawan** [comment]
I've added some tests and more guards on performing comparison of LocationSize. At least if in the future some issues are found, the error message is quite clear

**Review by davemgreen — COMMENTED** [review]
Is it worth adding a isScalable check to this line in Attributors getKnownNonNullAndDerefBytesForUse:
```
if (!Loc || Loc->Ptr != UseV || !Loc->Size.isPrecise() || I->isVolatile())
```

File: llvm/test/Transforms/LICM/memloc-vscale.ll [hunk 6]
```diff
@@ -0,0 +1,44 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
```
Comment (davemgreen) [inline llvm/test/Transforms/LICM/memloc-vscale.ll:1]: This test can be in llvm/test/Analysis/AliasSet

File: llvm/include/llvm/Analysis/MemoryLocation.h [hunk 7]
```diff
@@ -150,16 +159,22 @@ class LocationSize {
       return beforeOrAfterPointer();
     if (Value == AfterPointer || Other.Value == AfterPointer)
       return afterPointer();
+    if (isScalable() != Other.isScalable())
+      return afterPointer();
 
     return upperBound(std::max(getValue(), Other.getValue()));
```
Comment (davemgreen) [inline llvm/include/llvm/Analysis/MemoryLocation.h:165]: It looks like this max will implicitly convert to uint64_t. The != above for scalable vectors may need to be a ||.

File: llvm/test/Transforms/LICM/memloc-vscale.ll [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+; CHECK-LABEL: Alias sets for function 'ss':
+; CHECK: AliasSet[{{.*}}, 1] must alias, Mod       Pointers: (ptr %p, LocationSize::precise(vscale x 16))
+define void @ss(ptr %p) {
+  store <vscale x 2 x i64> zeroinitializer, ptr %p, align 2
+  store <vscale x 2 x i64> zeroinitializer, ptr %p, align 2
+  ret void
+}
+
```
Comment (davemgreen) [inline llvm/test/Transforms/LICM/memloc-vscale.ll:27]: This extra testcase for different sized scalable vectors was helpful for unionWith in AST
```
define void @ss2(ptr %p) {
  store <vscale x 2 x i64> zeroinitializer, ptr %p, align 2
  store <vscale x 2 x i64> zeroinitializer, ptr %p, align 2
  store <vscale x 4 x i64> zeroinitializer, ptr %p, align 2
  ret void
}
```

**Review by davemgreen — COMMENTED** [review]
I was looking through the uses of MemoryLocation and didn't notice any issues from what I could see, except for those above.

**Comment by nikic** [comment]
Looks like something went wrong here.

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp [hunk 9]
```diff
@@ -976,22 +977,30 @@ struct DSEState {
...
+    const TypeSize KillingSize = KillingLocSize.getValue();
+    const TypeSize DeadSize = DeadLoc.Size.getValue();
+    // Bail on doing Size comparison which depends on AA for now
+    // TODO: Remove AnyScalable once Alias Analysis deal with scalable vectors
+    const bool AnyScalable =
+        DeadSize.isScalable() || KillingLocSize.isScalable();
 
+    // TODO: Remove AnyScalable constraint once alias analysis fully support
+    // scalable quantities
+    if (AnyScalable)
```
Comment (davemgreen) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:989]: Is this change needed, if the code paths below already handle scalable sizes.

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp [hunk 10]
```diff
@@ -976,9 +977,17 @@ struct DSEState {
...
-    const uint64_t DeadSize = DeadLoc.Size.getValue();
+    const TypeSize KillingSize = KillingLocSize.getValue();
+    const TypeSize DeadSize = DeadLoc.Size.getValue();
+    // Bail on doing Size comparison which depends on AA for now
+    // TODO: Remove AnyScalable once Alias Analysis deal with scalable vectors
+    const bool AnyScalable =
+        DeadSize.isScalable() || KillingLocSize.isScalable();
 
+    // TODO: Remove AnyScalable constraint once alias analysis fully support
+    // scalable quantities
```
Comment (davemgreen) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:988]: This TODO is already above too.

**Review by davemgreen — COMMENTED** [review]
I've been looking through the uses of MemoryLocation and this couldn't see any other cases that immediately look like they would cause problems. I dont have a strong opinions on whether the DSE handling should be here or done as a separate patch. It probably makes sense to keep them separate. @nikic any thoughts?
